spec/onig-scanner-spec.js

```diff
@@ -0,0 +1,190 @@
+        let match = scanner.findNextMatchSync('ab…cde21', 5)
+        expect(match.index).toBe(1)
+
+        scanner = new OnigScanner(['\"'])
+        match = scanner.findNextMatchSync('{"…": 1}', 1)
+        expect(match.captureIndices).toEqual([{index: 0, start: 1, end: 2, length: 1}])
+      }))
+
+    describe('when the string searched contains surrogate pairs', () =>
+      it('counts paired characters as 2 characters in both arguments and return values', () => {
+        let scanner = new OnigScanner(['Y', 'X'])
+        let match = scanner.findNextMatchSync('a💻bYX', 0)
+        expect(match.captureIndices).toEqual([{index: 0, start: 4, end: 5, length: 1}])
+
+        match = scanner.findNextMatchSync('a💻bYX', 1)
+        expect(match.captureIndices).toEqual([{index: 0, start: 4, end: 5, length: 1}])
+
+        match = scanner.findNextMatchSync('a💻bYX', 3)
+        expect(match.captureIndices).toEqual([{index: 0, start: 4, end: 5, length: 1}])
+
+        match = scanner.findNextMatchSync('a💻bYX', 4)
+        expect(match.captureIndices).toEqual([{index: 0, start: 4, end: 5, length: 1}])
+
+        match = scanner.findNextMatchSync('a💻bYX', 5)
+        expect(match.index).toBe(1)
+        expect(match.captureIndices).toEqual([{index: 0, start: 5, end: 6, length: 1}])
+      }))
+
+    it("returns false when the input string isn't a string", () => {
+      let scanner = new OnigScanner(['1'])
+      expect(scanner.findNextMatchSync()).toBe(null)
+      expect(scanner.findNextMatchSync(null)).toBe(null)
+      expect(scanner.findNextMatchSync(2)).toBe(null)
+      expect(scanner.findNextMatchSync(false)).toBe(null)
+    })
+
+    it("uses 0 as the start position when the input start position isn't a valid number", () => {
+      let scanner = new OnigScanner(['1'])
+      expect(scanner.findNextMatchSync('a1', Infinity).index).toBe(0)
+      expect(scanner.findNextMatchSync('a1', -1).index).toBe(0)
+      expect(scanner.findNextMatchSync('a1', false).index).toBe(0)
+      expect(scanner.findNextMatchSync('a1', 'food').index).toBe(0)
+    })
+  })
+
+  describe('when the regular expression contains double byte characters', () =>
+    it('returns the correct match length', () => {
+      let scanner = new OnigScanner(['Возврат'])
+      let match = scanner.findNextMatchSync('Возврат long_var_name;', 0)
+      expect(match.captureIndices).toEqual([{index: 0, start: 0, end: 7, length: 7}])
+    }))
+
+  describe('when the input string contains invalid surrogate pairs', () =>
+    it('interprets them as a code point', () => {
+      let scanner = new OnigScanner(['X'])
+      let match = scanner.findNextMatchSync(`X${String.fromCharCode(0xd83c)}X`, 0)
+      expect(match.captureIndices).toEqual([{index: 0, start: 0, end: 1, length: 1}])
+
+      // Characters after unmatched high surrogates are not found.
+      match = scanner.findNextMatchSync(`X${String.fromCharCode(0xd83c)}X`, 1)
+      expect(match).toBeNull()
```

Comment:
This is the one behavior that's changed because we no longer transcode to UTF8. When dealing with invalid UTF16 (specifically, an unmatched high surrogate character), the unit that follows will no longer be considered as its own valid character, since it was supposed to contain the other half of the surrogate pair.